### PR TITLE
Return current directory without the trailing slash

### LINKED ISSUE #30891: dirname and @__DIR__ do not behave like bash, but pwd does
a trailing slash is ignored by bash but not julia.  intentional?

```
julia> dirname("/bar/foo")
"/bar"

julia> dirname("/bar/foo/")
"/bar/foo"

shell> dirname /bar/foo
/bar

shell> dirname /bar/foo/
/bar
```

relatedly, a trailing slash is added by julia, but not bash:

```
julia> @__DIR__
"/groups/scicompsoft/home/arthurb/projects/genie/screen/hek/src/"

shell> pwd
/groups/scicompsoft/home/arthurb/projects/genie/screen/hek/src
```

the above is all internally consistent in julia and kind of makes sense (dirs end in slash, files don't).  but then julia's `pwd` is not consistent (with `@__DIR__`) as it omits the trailing slash:

```
julia> pwd()
"/groups/scicompsoft/home/arthurb/projects/genie/screen"
```

why choose to behave like bash for `pwd` but not `dirname` or `@__DIR__` ?

`pwd` and `@__DIR__` both return directories, and so should both either add a trailing slash or not.

### PR DESCRIPTION
**Pull request by musm**
Fix https://github.com/JuliaLang/julia/issues/30891

### CODE REVIEW

File: base/path.jl [hunk 1]
```diff
@@ -145,7 +145,8 @@ end
 """
     dirname(path::AbstractString) -> AbstractString
 
-Get the directory part of a path.
+Get the directory part of a path. Trailing characters ('/' or '\\') in the path are
+counted as part of the path.
```
Comment (StefanKarpinski) [inline base/path.jl]: It would not be immediately clear to me what this meant, I don't think. Can you elaborate?

Comment (musm) [reply base/path.jl]: I wanted to clarify that unlike bash we do not ignore trailing slashes, e.g.
```julia
julia> dirname("/foo/bar/")
"/foo/bar"

julia> dirname("/foo/bar")
"/foo"
```

**Comment by musm** [comment]
It's not clear how one could test this.

File: base/path.jl (hunk 1 above)
Comment (StefanKarpinski) [reply base/path.jl]: Maybe write that and give an example?

**Comment by StefanKarpinski** [comment]
```jl
@test !endswith(@__DIR__, Base.Filesystem.path_separator_re)
@test !endswith(readchomp(`$(Base.julia_cmd()) -e 'print(@__DIR__)'`),
                Base.Filesystem.path_separator_re)
```

File: test/loading.jl [hunk 2]
```diff
@@ -43,6 +43,8 @@ include_string_test_func = include_string(@__MODULE__, "include_string_test() =
 @test basename(@__FILE__) == "loading.jl"
 @test isabspath(@__FILE__)
 
+@test !endswith(@__DIR__, Base.Filesystem.path_separator_re)
+@test !endswith(readchomp(`$(Base.julia_cmd()) -e 'print(@__DIR__)'`), Base.Filesystem.path_separator_re)
```
Comment (StefanKarpinski) [inline test/loading.jl]: I would put these inside the `let` block below and use `exename`.

File: test/loading.jl [hunk 3]
```diff
@@ -43,15 +43,17 @@ include_string_test_func = include_string(@__MODULE__, "include_string_test() =
...
 let exename = `$(Base.julia_cmd()) --compiled-modules=yes --startup-file=no`,
-    wd = sprint(show, abspath(pwd(), "")),
-    s_dir = sprint(show, joinpath(realpath(tempdir()), ""))
+    wd = sprint(show, pwd())
+    s_dir = sprint(show, realpath(tempdir()))
     @test wd != s_dir
     @test readchomp(`$exename -E "@__DIR__" -i`) == wd
     @test readchomp(`$exename -E "cd(()->eval(:(@__DIR__)), $s_dir)" -i`) == s_dir
     @test readchomp(`$exename -E "@__DIR__"`) == wd # non-interactive
+    @test !endswith(readchomp(`$exename -e "print(@__DIR__)"`), Base.Filesystem.path_separator)
```
Comment (StefanKarpinski) [inline test/loading.jl]: You could now remove this and test instead that `!endswith(wd, Base.Filesystem.path_separator)` and `!endswitht(s_dir, Base.Filesystem.path_separator)`.

File: test/loading.jl [hunk 4]
```diff
@@ -43,15 +43,17 @@ include_string_test_func = include_string(@__MODULE__, "include_string_test() =
 @test basename(@__FILE__) == "loading.jl"
 @test isabspath(@__FILE__)
 
+@test !endswith(@__DIR__, Base.Filesystem.path_separator)
```
Comment (StefanKarpinski) [inline test/loading.jl]: Seems weird to test this first. How about moving it down to line 49?

File: base/path.jl (hunk 1 above)
Comment (musm) [reply base/path.jl]: Better?

**Comment by musm** [comment]
good to merge ? 

@stevengj  maybe you wan't to take a look since you were involved in the original issue discussion

**Comment by StefanKarpinski** [comment]
Still needs NEWS.

**Comment by musm** [comment]
Yes, but there are already tests so can we remove that label

**Pull request merged**